tempesta_fw/http_limits.c

```diff
@@ -716,262 +674,319 @@ enum {
 
-#define __FRANG_FSM_MOVE(st)	T_FSM_MOVE(st, if (r) T_FSM_EXIT(); )
+#define __FRANG_FSM_MOVE(st)						\
+	T_FSM_MOVE(st, {						\
+	if (!r)								\
+		r = tfw_gfsm_move(&conn->state, (st), data);		\
+	if (r)								\
+		T_FSM_EXIT();						\
+	} )
 
-#define __FRANG_FSM_JUMP_EXIT(st)					\
+#define __FRANG_FSM_POSTPONE()						\
 do {									\
-	__fsm_const_state = st; /* optimized out to constant */		\
+	r = TFW_POSTPONE;						\
 	T_FSM_EXIT();							\
 } while (0)
 
+#define __FRANG_FSM_FINISH(ret)						\
+__fsm_done: __attribute__((unused))					\
+	T_DBG3("Finish %s FSM at state %#x, ret=%d\n",			\
+	       __fsm_name, __fsm_const_state, ret);
+
 static int
-frang_http_req_process(FrangAcc *ra, TfwConn *conn, TfwFsmData *data)
+frang_http_req_hdrs_check(FrangAcc *ra, TfwConn *conn,
+			  TfwFsmData *data)
 {
 	int r = TFW_PASS;
+	TfwHttpReq *req = (TfwHttpReq *)data->req;
+	__FRANG_CFG_VAR(field_len, http_field_len);
+
+	/* Ensure that singular header fields are not duplicated. */
+	if (test_bit(TFW_HTTP_B_FIELD_DUPENTRY, req->flags)) {
+		frang_msg("duplicate header field found",
+			  &FRANG_ACC2CLI(ra)->addr, "\n");
+		goto block;
+	}
+
+	/* Ensure that length of all parsed headers fields is within limits. */
+	if (field_len && (r = frang_http_field_len(req, ra, field_len)))
+		goto block;
+
+	return TFW_PASS;
+block:
+	return TFW_BLOCK;
+}
+
+static int
+frang_http_req_incomplete_hdrs_check(FrangAcc *ra, TfwConn *conn,
+				     TfwFsmData *data)
+{
 	TfwHttpReq *req = (TfwHttpReq *)data->req;
 	struct sk_buff *skb = data->skb;
 	struct sk_buff *head_skb = req->msg.skb_head;
 	__FRANG_CFG_VAR(hdr_tmt, clnt_hdr_timeout);
 	__FRANG_CFG_VAR(hchnk_cnt, http_hchunk_cnt);
-	T_FSM_INIT(Frang_Req_0, "frang");
 
-	BUG_ON(!ra);
-	BUG_ON(req != container_of(conn->msg, TfwHttpReq, msg));
-	frang_dbg("check request for client %s, acc=%p\n",
+	frang_dbg("check incomplete request headers for client %s, acc=%p\n",
 		  &FRANG_ACC2CLI(ra)->addr, ra);
-
-	spin_lock(&ra->lock);
-
 	/*
 	 * There's no need to check for header timeout if this is the very
 	 * first chunk of a request (first full separate SKB with data).
 	 * The FSM is guaranteed to go through the initial states and then
 	 * either block or move to one of header states. Then header timeout
 	 * is checked on each consecutive SKB with data - while we're still
 	 * in one of header processing states.
-	 *
-	 * Why is this not one of FSM states? Basically, that's to avoid
-	 * going through unnecessary FSM states each time this is run. When
-	 * there's a slowris attack, we may stay long in Hdr_Method or in
-	 * Hdr_UriLen states, and that would require including the header
-	 * timeout state in the loop. But when we're past these states, we
-	 * don't want to run through them on each run again, and just want
-	 * to loop in FieldDup and FieldLen states. I guess that can be
-	 * done with some clever FSM programming, but this is just simpler.
```

Comment:
OK, that absolutely makes sense to move the checks and call them only when necessary.